osquery/tables/system/windows/objects.cpp

```diff
@@ -0,0 +1,229 @@
+//
+// example at:
+//    http://pastebin.com/embed_js/zhmJTffK
+//    https://randomsourcecode.wordpress.com/2015/03/14/enumerating-deviceobjects-from-user-mode/
+//    https://msdn.microsoft.com/en-us/library/bb470238(v=vs.85).aspx
+//
+Status enumerateObjectNamespace(const std::wstring& directory,
+                                std::vector<obj_name_type_pair>& objects) {
+  // look up addresses of NtQueryDirectoryObject and
+  // NtQuerySymbolicLinkObject.  Both are exported from ntdll
+  //
+  // NtQueryDirectoryObject is documented on MSDN, there is no
+  // associated header or import library
+  //
+  auto NtQueryDirectoryObject =
+      (NTQUERYDIRECTORYOBJECT)GetProcAddress(GetModuleHandleA("ntdll"),
+                                             "NtQueryDirectoryObject");
+  if (NULL == NtQueryDirectoryObject) {
+    return Status(GetLastError(), "Unable to find NtQueryDirectoryObject");
+  }
+
+  // open the caller-provided root directory
+  // kdo will manage the resultant HANDLE, which is also used to query
+  // for object name and object type name pairs below
+  //
+  Handle kdo;
+  auto status = kdo.openDirObj(directory);
+  if (!status.ok()) {
+    return status;
+  }
+
+  // iterator index is incremented by NtQueryDirectoryObject
+  // for safety, bail out at max_supported_objects 
+  //
+  for (unsigned long index = 0; index < max_supported_objects;) {
+    unsigned char obj_buf[obj_buf_size] = {0};
+    auto pObjDirInfo = reinterpret_cast<POBJDIR_INFORMATION>(obj_buf);
+
+    //  get the name and type of the index'th object in the directory
+    //
+    auto ntStatus = NtQueryDirectoryObject(kdo.getAsHandle(),
+                                               pObjDirInfo,
+                                               obj_buf_size,
+                                               TRUE,
+                                               FALSE,
+                                               &index,
+                                               NULL);
+    if (STATUS_SUCCESS != ntStatus) {
+      break;
+    }
+
+    obj_name_type_pair object;
+    object.first = (pObjDirInfo->ObjectName.Buffer);
+    object.second = (pObjDirInfo->ObjectTypeName.Buffer);
+
+    objects.push_back(object);
+  }
+
+  return Status();
+}
+
+// enumerate all objects in a given windows terminal services session
+//
+// objects are found in the windows object directory
+// "\Sessions\BNOLINKS\<sessionnum>"
+//
+Status enumerateBaseNamedObjectsLinks(const std::wstring& session_num,
+                                      const std::wstring& object_type,
+                                      std::vector<obj_name_type_pair>& objects) {
+  // look up NtQuerySymbolicLinkObject as exported from ntdll
+  //
+  auto NtQuerySymbolicLinkObject =
+      (NTQUERYSYMBOLICLINKOBJECT)GetProcAddress(GetModuleHandleA("ntdll"),
+                                                "NtQuerySymbolicLinkObject");
+  if (NULL == NtQuerySymbolicLinkObject) {
```

Comment:
It is much better to use `nullptr` instead of `NULL`, [some explanation](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es47-use-nullptr-rather-than-0-or-null)

Reply:
no problem

Reply:
done